HW2/P2/parallel_vector.pyx

```diff
@@ -65,51 +65,96 @@ cpdef move_data_serial(np.int32_t[:] counts,
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+#   print("Number of locks: {}".format(counts.shape[0]))
+   for r in range(repeat):
+       for idx in prange(src.shape[0], 
+                         nogil=True, 
+                         num_threads=4):
+            src_loc = src[idx]
+            dest_loc = dest[idx]
+
+            # Perform a useless computation here to match the computation 
+            # of the medium grained version
+            lock_num_src = (src_loc / N)
+            lock_num_dest = (dest_loc / N)
+
+            # If source and destination are the same, 
+            # there's nothing to do.
+            if src_loc != dest_loc: 
+                # To prevent deadlock, let's lock from the lower index
+                # This costs some overhead
+                if src_loc < dest_loc:
+                    acquire(&(locks[src_loc]))
+                    acquire(&(locks[dest_loc]))
+                elif src_loc > dest_loc:
+                    acquire(&(locks[dest_loc]))
+                    acquire(&(locks[src_loc]))
+                if counts[src_loc] > 0:
+                    counts[src_loc] -= 1
+                    counts[dest_loc] += 1
+                release(&(locks[src_loc]))
+                release(&(locks[dest_loc]))
 
    free_N_locks(counts.shape[0], locks)
 
-
+############# MEDIUM GRAINED #############
 cpdef move_data_medium_grained(np.int32_t[:] counts,
                                np.int32_t[:] src,
                                np.int32_t[:] dest,
                                int repeat,
                                int N):
    cdef:
-       int idx, r
+       int idx, r, src_loc, dest_loc, lock_num_src, lock_num_dest
        int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
        omp_lock_t *locks = get_N_locks(num_locks)
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+#   print("Number of locks: {}".format(num_locks))
+
+        
+   for r in range(repeat):
+       for idx in prange(src.shape[0], 
+                         nogil=True, 
+                         num_threads=4):
+            src_loc = src[idx]
+            dest_loc = dest[idx]
+
+            lock_num_src = (src_loc / N)
```

Comment:
what if these values change (e.g. src[idx] changes between now and acquiring the lock?)